client/flutter/lib/pages/onboarding/location_sharing_page.dart

```diff
@@ -35,16 +35,52 @@ class _LocationSharingPageState extends State<LocationSharingPage> {
       if (await Location().hasPermission() == PermissionStatus.granted) {
         if (await Location().requestService()) {
           LocationData location = await Location().getLocation();
-          Map jitteredLocationData = JitterLocation().jitter(
+          final geo = Geolocator();
+          // Use en_US because these are not for display, they are for indexing.
+          final myPlaces = await geo.placemarkFromCoordinates(
               location.latitude, location.longitude,
-              5 /*kms refers to kilometers*/);
+              localeIdentifier: 'en_US');
+          if (myPlaces.isEmpty) {
+            print('No Reverse Geolocation place');
+            return;
+          }
+          final myPlace = myPlaces.first;
+          if (myPlace.isoCountryCode == null ||
+              myPlace.isoCountryCode.isEmpty) {
+            print('No Reverse Geolocation country');
+            // We need at least a country.
+            return;
+          }
+          var addrComponents = [myPlace.isoCountryCode];
+          [
+            myPlace.administrativeArea,
+            myPlace.subAdministrativeArea,
+            myPlace.locality,
+          ].forEach((n) {
+            if (n != null && n.isNotEmpty) {
+              addrComponents.insert(0, n);
+            }
+          });
+          final addr = addrComponents.join(', ');
+
+          final placesCityCenter =
+              await geo.placemarkFromAddress(addr, localeIdentifier: 'en_US');
+          if (placesCityCenter.isEmpty) {
+            print('No Geolocated City Center');
+            return;
+          }
+          final placeCityCenter = placesCityCenter.first;
 
           await WhoService.putLocation(
-              latitude: jitteredLocationData['lat'],
-              longitude: jitteredLocationData['lng']);
+              latitude: placeCityCenter.position.latitude,
+              longitude: placeCityCenter.position.longitude,
+              countryCode: myPlace.isoCountryCode,
+              adminArea: myPlace.administrativeArea,
+              subadminArea: myPlace.subAdministrativeArea,
+              locality: myPlace.locality);
         }
       }
-    } catch(_) {
+    } catch (_) {
       // ignore for now.
```

Comment:
Can we log errors using Google Analytics? I have concerns that Apple geocoder and reverse geocoder may not be as comprehensive at Google's, particularly in developing countries. I would at least like to be able to measure problems like that so they don't silently fail. If the problem is widespread in certain areas, we might look at using Google's Geocoding API so that we have consistency across both platforms.

https://developers.google.com/maps/documentation/geocoding/intro

Reply:
Thanks Bruno I agree about possibly using Google Maps on both platforms - we’ll need to use api keys with a quota if we don’t use the platform provider, so let’s not do it tonight. I’ve made a bug #876 to log all errors in analytics (it’s a bigger problem than just location APIs, and we need to think though how to ensure no exception text PII gets sent to google analytics).

However, on the issue of geocoders’ lack of comprehensiveness, we will not see that as errors. Instead we will see it as a partially incomplete location stored on the server (lacking city, county, etc. but containing country). We cannot distinguish between genuine lack of detail (such as unincorporated areas of US counties) and the geocoder’s inability to discern such detail. We could however run a statistical test comparing iOS to Google Maps later if needed.

Reply:
Country code is the main thing. Not sure when WHO will be ready for even country specific notifications.